src/main/scala/cognite/spark/v1/DefaultSource.scala

```diff
@@ -85,13 +91,67 @@ class DefaultSource
     new SequenceRowsRelation(config, sequenceId)(sqlContext)
   }
 
+  private def getFunctionResult(client: GenericClient[IO], functionId: Long, result: FunctionCall): IO[Json] = {
+    var res = result
+    while (res.status != None && res.status.get == "Running") {
+      res = client
+        .functionCalls(functionId)
+        .retrieveById(res.id.get)
+        .unsafeRunSync()
+    }
+
+    if (res.status != None && res.status.get == "Completed") {
+      for {
+        functionCallResponse <- client
+          .functionCalls(functionId)
+          .retrieveResponse(res.id.getOrElse(0L))
+      } yield functionCallResponse.response.getOrElse(Json.fromJsonObject(JsonObject.empty))
+    } else {
+      IO.pure(Json.fromJsonObject(JsonObject.empty))
+    }
+  }
+
+  private def callFunctionAndGetResult(config: RelationConfig, functionId: Long, data: Json): IO[Json] = {
+    implicit val backend: SttpBackend[IO, Nothing] = CdpConnector.retryingSttpBackend(config.maxRetries)
+    for {
+      client <- GenericClient.forAuth[IO](Constants.SparkDatasourceVersion, config.functionsAuth.get, config.baseUrl, apiVersion = Some("playground"))
+      call <- client.functionCalls(functionId).callFunction(Json.fromJsonObject(JsonObject.empty))
+      res <- getFunctionResult(client, functionId, call)
+    } yield res
+  }
+
+  private def initializeFunctionUdfs(config: RelationConfig, context: SQLContext): Unit = {
+    implicit val backend: SttpBackend[IO, Nothing] = CdpConnector.retryingSttpBackend(config.maxRetries)
+
+    val functionsIO = for {
+      client <- GenericClient.forAuth[IO](Constants.SparkDatasourceVersion, config.functionsAuth.get, config.baseUrl, apiVersion = Some("playground"))
+      functions <- client.functions.read()
+      l <- client.login.status()
+      f = functions.items.map(func =>
+        (func.name.replace(" ", "_"), (data: Any) =>
```

Comment:
For args, we can use `data: Row`. The user can then call the function as `cdf_function("foo", struct(*))` to pass all columns. Our UDF will receive the columns in a `Row` object which also includes a schema. The schema doesn't matter too much as we should just call `to_json` (that's what it's called in SQL, there should be something similar in the Scala API) to encode it as a string.

For the return type we have two options that I can see, either we make it to always be a string, and let the user decide if they should decode it as JSON, or we make it a `Map[String, String]` where the values would be JSONs encoded as strings. For simple cases the latter is a bit more convenient as the user can do `cast(cdf_function("my function", struct(*)).prediction as double)` and similar, which is probably a common use case.

Reply:
If functions are allowed to return bare numbers we'd have to go for always returning a string, though. We can add some additional methods later like `cdf_function_double` which would return a double, or something, but that's not needed for the first release.

Reply:
I could not figure out how to convert a row into JSON. The `to_json` function accepts a `Column` type. All internet examples I found used it from SQL. So I think it is best if the user uses `to_json` from SQL.